### PR TITLE
qe: Correctly handle rollback on mssql when transaction have been automatically rolled back

### PR DESCRIPTION
**Pull request by SevInf**
Tiberius PR: https://github.com/prisma/tiberius/pull/245
Quaint PR: https://github.com/prisma/quaint/pull/405

After fixes in tiberius in quaint, manual rollback after transaction
have been closed automatically because of error will fail with a
different message/code. This commit adapts the engine to the changes.

Part of: https://github.com/prisma/prisma/issues/15607

### CODE REVIEW

File: migration-engine/qe-setup/src/lib.rs [hunk 1]
```diff
@@ -87,12 +87,12 @@ pub async fn teardown(prisma_schema: &str, db_schemas: &[&str]) -> ConnectorResu
 
 #[derive(Default)]
 struct LoggingHost {
-    printed: std::sync::Mutex<Vec<String>>,
+    printed: parking_lot::Mutex<Vec<String>>,
```
Comment (pimeys) [inline migration-engine/qe-setup/src/lib.rs:90]: Even better would be `tokio::sync::Mutex`, but I didn't want to change any migration engine apis.

Comment (pimeys) [reply migration-engine/qe-setup/src/lib.rs:90]: Standard `MutexGuard` is `!Send`, and is nasty to work with.

**Review by garrensmith — APPROVED** [review]
Nice work. You basically had to re-implement itx to get this working. We can use this as a base for making it easier to test concurrency in the engine.

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        let tag = ConnectorTag::try_from(("sqlserver", None))?;
+
+        let datamodel = render_test_datamodel(
+            &*CONFIG,
+            "sql_server_deadlocks_test",
+            SCHEMA.to_owned(),
+            &[],
+            None,
+            &[],
+            Some("READ COMMITTED"),
```
[the remark this answers is not in the document]
Comment (pimeys) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs:73]: This is the lowest level of isolation that deadlocks. We use `READ UNCOMMITTED` in most of the query engine tests to not deadlock when we modify the information schema. Did not try out other isolation levels due to this PR not really being about that, but about error recovery after a deadlock.

**Comment by pimeys** [comment]
> Nice work. You basically had to re-implement itx to get this working. We can use this as a base for making it easier to test concurrency in the engine.

We have a multi-engine setup in the migration tests. You basically have a test api you can use to spawn engines. Even better here would be if we'd have a sync engine actor, and we can just write normal tests without async to make the compilation much faster.

File: query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs [hunk 3]
```diff
@@ -37,6 +37,7 @@ pub fn render_test_datamodel(
     excluded_features: &[&str],
     relation_mode_override: Option<String>,
     db_schemas: &[&str],
+    isolation_level: Option<&'static str>,
```
Comment (tomhoule) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs:40]: Does the isolation need to be in the connection string? Could it be set after the connection is established, in the connection? Ultimately, I think we want to stop constructing connection strings in query engine test setup code

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    RollbackTransaction(TxId),
+    SetActiveTx(TxId),
+}
+
+#[derive(Debug)]
+enum Response {
+    Query(TestResult<QueryResult>),
+    Tx(TestResult<TxId>),
+    Rollback(Result<(), user_facing_errors::Error>),
+}
```
Comment (tomhoule) [inline query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs:51]: one alternative pattern I like is sending closures through the channel. But this is also good, for code that has to be maintained :+1:

File: Cargo.lock [hunk 5]
```diff
@@ -3065,6 +3065,7 @@ dependencies = [
  "migration-core",
  "mongodb",
  "mongodb-client",
+ "parking_lot 0.12.1",
```
Comment (tomhoule) [inline Cargo.lock:3068]: I think it would be worth digging into why we have multiple versions of parking lot in Cargo.lock

**Review by tomhoule — APPROVED** [review]
nice! I agree with @garrensmith that this should be used as a base or inspiration for better query engine tests, and then made uniform. Maintaining that setup for a single test may become a lot of work.

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs (hunk 4 above)
Comment (pimeys) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/regressions/prisma_15607.rs:51]: More like erlang style...

File: query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs (hunk 3 above)
Comment (pimeys) [reply query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs:40]: Yeah, that would be really nice. There is a mechanism in quaint that sets the isolation level, if it's easy enough I'd like to use that. If not, we'll go with this.

File: Cargo.lock (hunk 5 above)
Comment (pimeys) [reply Cargo.lock:3068]: metrics-exporter-prometheus uses 0.11...

**Pull request merged**